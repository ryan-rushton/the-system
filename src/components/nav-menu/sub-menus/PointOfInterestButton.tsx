import clsx from 'clsx';
import { FC } from 'react';
import { useTranslation } from 'react-i18next';
import styles from './PointOfInterestButton.module.scss';

interface Props {
  /** Whether the component is currently visible. */
  isVisible: boolean;
  /** Point of interest object. */
  pointOfInterest: { id: string };
  /** Whether the point of interest is currently being followed on screen. */
  isBeingFollowed: boolean;
  /** Click handler for when the button is clicked. */
  onPoiClick(): void;
}

/**
 * A button for points of interest. When Clicked the screen will follow the poi until
 * the button is clicked again or another poi is clicked.
 */
const PointOfInterestButton: FC<Props> = ({ isVisible, pointOfInterest, isBeingFollowed, onPoiClick }) => {
  const { t } = useTranslation();

  const displayName = t(`pointsOfInterest.${pointOfInterest.id}`);
  return (
    <button
      key={pointOfInterest.id}
      data-testid={`${pointOfInterest.id}-button`}
      className={clsx(styles.navItem, { [styles.navItemFollowed]: isBeingFollowed })}
      onClick={onPoiClick}
      tabIndex={isVisible ? 0 : undefined}
      aria-label={displayName}
    >
      <span>{displayName}</span>
    </button>
  );
};

export default PointOfInterestButton;
