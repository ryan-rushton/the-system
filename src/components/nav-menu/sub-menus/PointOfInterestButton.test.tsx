import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import PointOfInterestButton from './PointOfInterestButton';

describe('PointOfInterestButton', () => {
  afterEach(cleanup);

  test('it renders', () => {
    const poi = { id: 'somePoint', ref: React.createRef() };
    render(
      <PointOfInterestButton pointOfInterest={poi} isVisible={true} isBeingFollowed={false} onPoiClick={jest.fn()} />
    );
    const element = screen.getByText(/^Some point$/);

    expect(element).toBeInTheDocument();
  });

  test('it matches snapshot', () => {
    const poi = { id: 'somePoint', ref: React.createRef() };
    render(
      <PointOfInterestButton pointOfInterest={poi} isVisible={true} isBeingFollowed={false} onPoiClick={jest.fn()} />
    );

    expect(screen.getByTestId('somePoint-button')).toMatchSnapshot();
  });

  test('it changes orbits visible when Show Orbits clicked', () => {
    const poi = { id: 'somePoint', ref: React.createRef() };
    const onPoiClick = jest.fn();

    render(
      <PointOfInterestButton pointOfInterest={poi} isVisible={true} isBeingFollowed={false} onPoiClick={onPoiClick} />
    );

    userEvent.click(screen.getByText(/^Some point$/));

    expect(onPoiClick).toBeCalledTimes(1);
  });

  test('it changes orbits visible when Show Orbits enter pressed', () => {
    const poi = { id: 'somePoint', ref: React.createRef() };
    const onPoiClick = jest.fn();

    render(
      <PointOfInterestButton pointOfInterest={poi} isVisible={true} isBeingFollowed={false} onPoiClick={onPoiClick} />
    );

    fireEvent.keyDown(screen.getByText(/^Some point$/), { key: 'Enter', keyCode: 13 });

    expect(onPoiClick).toBeCalledTimes(1);
  });
});
