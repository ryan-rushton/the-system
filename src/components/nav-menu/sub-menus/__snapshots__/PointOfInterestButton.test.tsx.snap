// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PointOfInterestButton > it matches snapshot 1`] = `
<button
  aria-label="Some point"
  class="_navItem_e88351 _button_fb7df1"
  data-testid="somePoint-button"
  tabindex="0"
>
  <span>
    Some point
  </span>
</button>
`;
